### PR TITLE
fix(gnovm): correct doOpDefer value-stack handling for defer f(g())

### PR DESCRIPTION
**Pull request by omarsy**
## Description

`defer f(g())`, where the single syntactic argument is a multi-value call (`g` returns exactly the arguments `f` expects — valid, idiomatic Go), was mis-handled by the VM.

`doOpDefer` used `len(ds.Call.Args)` (which is `1` for the `f(g())` spread form) instead of `ds.Call.NumArgs` (the expanded argument-*value* count) for both the func-locating `PeekValue` offset and the `popCopyArgs` count. It therefore peeked an *argument* value as the function and popped the wrong number of values:

- `defer h(twovals())` — `h(a, b int)`, `twovals() -> (7, 8)` — panicked with `runtime error: defer called a nil function` instead of calling `h`.
- `defer f(g())` where `g()`'s first result is a func value — ran the **wrong** function (that result value) and skipped `f` entirely, silently, with no panic.

Both were verified against `go run` of the equivalent Go program.

## Fix

- Use `ds.Call.NumArgs` for the offset and pop count. This mirrors the normal call path (`doOpCall` already uses `PeekValue(1 + cx.NumArgs)`). `NumArgs == len(Args)` for every non-spread call, so ordinary defers are unaffected; it differs only for the `x(f())` spread form, which is exactly the broken case.
- Fix the `case nil:` branch (deferring a nil func value with arguments): it popped only the func value and left the evaluated arguments on the value stack. The leak was latent (frame truncation on return discards it) but violated the stack-neutral invariant the other branches maintain and would grow the stack in a loop. It now discards the arguments so every branch balances.

## Tests

New regression filetests under `gnovm/tests/files/`:
- `defer_multivalue_arg.gno` — the two headline cases; panics/misbehaves on the pre-fix VM, passes after.
- `defer_nil_func_args.gno` — the nil-func-with-args branch.

Verification suite (per `AGENTS.md`), all green:
- `go test ./gnovm/pkg/gnolang/ -run Files -test.short`
- `go test ./gno.land/pkg/sdk/vm/ -run Gas`
- `go test ./gno.land/pkg/integration/ -run TestTestdata`
- `cd examples && go run ../gnovm/cmd/gno test ./...`

## ADR

`gnovm/adr/pr6137_defer_multivalue_arg_desync.md`.

## AI assistance

Developed with AI assistance (Claude Code). The diagnosis and fix were verified against `go run` and independently reviewed; authored and reviewed by @omarsy.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: omarsy/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🟢 User davd-gzl already reviewed PR 6137 with state APPROVED
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: gnovm/tests/files/defer_multivalue_arg.gno [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+func twovals() (int, int) { return 7, 8 }
+func h(a, b int)          { println("h:", a+b) }
+
+func marker()            { println("WRONG: marker ran") }
+func g() (func(), int)   { return marker, 99 }
+func f(fn func(), n int) { println("f: n =", n) }
+
+func main() {
+	defer f(g())       // runs 2nd: must print "f: n = 99", must not run marker
```
Comment (davd-gzl) [inline gnovm/tests/files/defer_multivalue_arg.gno:18]: Missing test: `defer called a nil function` also hit a variadic callee, a bound method and a builtin at the base, and none is covered.

<details><summary>test cases</summary>

Green at 861df2c42, and at the merge base with `op_call.go` reverted each of the three fails with `unexpected panic: runtime error: defer called a nil function`.

```go
package main

type counter struct{ n int }

func (c counter) add(a, b int) { println("method:", c.n+a+b) }

func two() (int, int) { return 7, 8 }

func variadic(a int, rest ...int) { println("variadic:", a, len(rest), rest[0]) }

func main() {
	defer println(two())           // runs 3rd: builtin callee
	defer counter{n: 1}.add(two()) // runs 2nd: bound method callee
	defer variadic(two())          // runs 1st: variadic callee
	println("body")
}

// Output:
// body
// variadic: 7 1 8
// method: 16
// 7 8
```
</details>

File: gnovm/pkg/gnolang/op_call.go [hunk 2]
```diff
@@ -739,6 +744,10 @@ func (m *Machine) doOpDefer() {
 		cfr.PushDefer(Defer{Callable: cv, Args: args, Source: ds})
 	case nil:
 		// deferred a nil func value; raised as call-of-nil at the deferred call.
+		// The other branches consume the argument values via popCopyArgs; do
+		// the same here (discarding them) so the value stack stays balanced.
+		// The func value itself is popped below.
+		m.PopValues(numArgs)
 		cfr.PushDefer(Defer{Source: ds})
 	default:
 		m.pushPanic(typedString(fmt.Sprintf("invalid defer function call: %v", cv)))
```
Comment (davd-gzl) [inline gnovm/pkg/gnolang/op_call.go:753]: Suggestion: `default:` returns before the trailing [`m.PopValue()`](https://github.com/gnolang/gno/blob/861df2c42/gnovm/pkg/gnolang/op_call.go#L756), so it leaves the func value and its `numArgs` arguments on the value stack.

```suggestion
	default:
		m.PopValues(numArgs + 1)
		m.pushPanic(typedString(fmt.Sprintf("invalid defer function call: %v", cv)))
```

<details><summary>what it changes</summary>

Nothing reachable: `defer int64(x)` is the only shape found to land in `default:`, and the go typechecker rejects it with `defer requires function call, not conversion int64(x) (value of type int64)`. With the suggestion applied a filetest on that source still reports `invalid defer function call: typeval{int64}` and the same typecheck error, and the 35 filetests matching `TestFiles/defer` stay green.
</details>

**Review by davd-gzl — COMMENTED** [review]
> AI review, claude-opus-5 at xhigh, [skills](https://github.com/davd-gzl/skills) · Status: APPROVE

**Review by davd-gzl — APPROVED** [review]

**Review by ltzmaxwell — CHANGES_REQUESTED** [review]
Thanks for the clean diagnosis — the `NumArgs` vs `len(Args)` analysis is right. Two things changed under this PR while it was open, though, so I think it needs a reshape before it can land.

### 1. Master already carries the core fix

`doOpDefer` on master now reads `numArgs := ds.Call.NumArgs` — it landed as part of #6164 (0099b3138), together with `gnovm/tests/files/defer_multivalue_call.gno`, which covers both the scalar-spread and func-typed-first-result cases this PR targets. As a result this branch now conflicts in `op_call.go` (`git merge-tree origin/master HEAD`), and after a merge the `NumArgs` hunk and `defer_multivalue_arg.gno` are duplicates of what's already there.

What this PR still uniquely adds is the `case nil:` `PopValues(numArgs)` so every branch of `doOpDefer` is value-stack-neutral. That's a valid invariant to enforce, and I agree with @davd-gzl's second comment that the `default:` branch should get the same treatment (`PopValues(numArgs + 1)`).

### 2. `defer_nil_func_args.gno` doesn't test the change

With the new `m.PopValues(numArgs)` line deleted, this filetest still passes. Frame truncation on return discards the leaked slots, so nothing visible to Gno code changes; what the file does assert — deferring a nil func panics and is recoverable — is already covered by `func31.gno` and `recover26.gno`.

On master the leak *is* observable, because GC now recounts the operand stack. Measured with three `defer nf(1, 2)` followed by `runtime.GC(); println(runtime.MemStats())` under `// MAXALLOC: 100000000`:

| | `bytes` |
|---|---|
| without `PopValues` | 7856 |
| with `PopValues` | 7616 |

240 B = 6 leaked `TypedValue` slots × 40 B. So a MemStats golden in the style of `alloc_defer_gc.gno` pins the invariant and fails without the pop — that's the test I'd suggest replacing `defer_nil_func_args.gno` with.

The same pin covers the `default:` branch, because the filetest harness records the Go typechecker rejection under `// TypeCheckError:` and still runs the VM. `defer int64(x)` recovered in a deferred closure that calls `runtime.GC()` and prints `runtime.MemStats()`:

| | `bytes` |
|---|---|
| without `PopValues(numArgs + 1)` | 9299 |
| with `PopValues(numArgs + 1)` | 9219 |

80 B = the leftover func value and its one argument. The deferred closures run before the frame is truncated (`pushPanic` → `PopUntilLastCallFrame` leaves `m.Values` alone), which is why the leftovers are still counted there. One filetest with both shapes would pin both branches.

### Suggested path

Merge master, drop the duplicated `NumArgs` hunk and `defer_multivalue_arg.gno`, keep the nil-branch pop plus the `default:` pop, replace the nil-branch filetest with the MemStats pin, and trim the ADR to the stack-neutrality change. Alternatively, if you'd rather not carry a tidy-up-only PR, closing as superseded by #6164 is fine too.

Requesting changes for the merge conflict and the test; the code itself is correct.
